Sources/NIOCore/NIOIPAddresses.swift

```diff
@@ -0,0 +1,583 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+
+import Darwin
+import CNIOLinux
+import CoreFoundation
+
+
+/// Represent a single `NIOIPAddress`.
+public enum NIOIPAddress: CustomStringConvertible {
+    /// `Error` that may be thrown if we fail to create a `NIOIPAddress`
+    public enum IPAddressError: Error {
+        /// Given string input is not supported IP Address
+        case failedToParseIPString(String)
+        /// Given string input is not supported IP Address
+        case bytesArrayHasWrongLength(Int)
+    }
```

Comment:
I think it's reasonable to provide errors: they can always be converted to `nil` when users don't care about the information, but they can't get that information back from a `nil` later.

Reply:
It does cost, though, both in binary size for the extra type and its conformances and in code-paths which cannot be merged because they have to throw different errors (one of which containing an ARC-managed String).

In my own parser, I've seen reasonable gains by specialising errors away via generics, and if I were writing it again, I wouldn't include them at all. How is anybody going to handle the error that the bytes array had the wrong length, if they got it wrong the first time? Nobody guesses when it comes to the bytes length - either it's correct or something irrecoverably bad happened.

Reply:
Yeah that's what I thought, too - but then I figured all callers would need to handle the failure either way, and actually, the more common thing is for people to parse hostnames first as an IP address and treat it as a domain as a fallback. So they'll never even see the detailed error. And if there is no fallback, the caller will need to handle it in some other way, perhaps by throwing their own "invalid IP address <somestring>" or "invalid hostname <somestring>". But it's their choice and adding an error type doesn't actually add anything.

It's analogous to `Int(String)`. It doesn't bother throwing detailed errors; it isn't worth it, and there's not realistically anything you can do to handle it, so it's better to be a failible initialiser.

That's not to mention that it's a public enum, so it couldn't evolve to be any more detailed than the current 2 cases.

Reply:
I don't disagree that those use-cases exist: I'm just leaning much more strongly towards having more diagnostic information and letting users throw it away, not less. The trade off between failable initialisers and throwing ones is a tight space, and I definitely think this is in a "gut feel" situation. If we find it's totally not worth it, we can always deprecate and replace the API.

Reply:
It's totally up to you and I'm not arguing with it, but I just wanted to add that this:

> I'm just leaning much more strongly towards having more diagnostic information and letting users throw it away, not less

Is what I used to think as well. But it turns out, at scale across complex parsers with tight performance requirements, there can be a significant cost to that because it means the compiler effectively can't merge loops and perform other optimisations; it's tightly bound to your code pretty-much as you wrote it.

Then when you add the additional types, conformances and whatnot, the awkwardness in building an interface that can evolve... it totally turned me in the other direction. Add errors where you really expect they might meaningfully be handled or used in user-facing diagnostics, or you need some clearer indication at the call-site of a failable operation. That's my approach these days.